drivers/modem/quectel-bg9x.c

```diff
@@ -0,0 +1,1155 @@
+/*
+ * Copyright (c) 2020 Analog Life LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT quectel_bg9x
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(modem_quectel_bg9x, CONFIG_MODEM_LOG_LEVEL);
+
+#include "quectel-bg9x.h"
+
+static struct k_thread	       modem_rx_thread;
+static struct k_work_q	       modem_workq;
+static struct modem_data       mdata;
+static struct modem_context    mctx;
+static const struct socket_op_vtable offload_socket_fd_op_vtable;
+
+K_KERNEL_STACK_DEFINE(modem_rx_stack, CONFIG_MODEM_QUECTEL_BG9X_RX_STACK_SIZE);
+K_KERNEL_STACK_DEFINE(modem_workq_stack, CONFIG_MODEM_QUECTEL_BG9X_RX_WORKQ_STACK_SIZE);
+NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF, MDM_RECV_BUF_SIZE, 0, NULL);
+
+static int net_offload_dummy_get(sa_family_t family, enum net_sock_type type,
+				 enum net_ip_protocol ip_proto,
+				 struct net_context **context)
+{
+
+	LOG_ERR("CONFIG_NET_SOCKETS_OFFLOAD must be enabled for this driver");
+	return -ENOTSUP;
+}
+
+static inline uint8_t *modem_get_mac(const struct device *dev)
+{
+	struct modem_data *data = dev->data;
+	uint32_t hash_value;
+
+	data->mac_addr[0] = 0x00;
+	data->mac_addr[1] = 0x10;
+
+	/* use IMEI for mac_addr */
+	hash_value = hash32(mdata.mdm_imei, strlen(mdata.mdm_imei));
+
+	UNALIGNED_PUT(hash_value, (uint32_t *)(data->mac_addr + 2));
+
+	return data->mac_addr;
+}
+
+/* Func: modem_atoi
+ * Desc: Convert string to long integer, but handle errors
+ */
+static int modem_atoi(const char *s, const int err_value,
+		      const char *desc, const char *func)
+{
+	int   ret;
+	char  *endptr;
+
+	ret = (int)strtol(s, &endptr, 10);
+	if (!endptr || *endptr != '\0') {
+		LOG_ERR("bad %s '%s' in %s", log_strdup(s), log_strdup(desc),
+			log_strdup(func));
+		return err_value;
+	}
+
+	return ret;
+}
+
+/* Func: modem_at
+ * Desc: Send "AT" command to the modem and wait for it to
+ * respond. If the modem doesn't respond after some time, give
+ * up and kill the driver.
+ */
+static int modem_at(struct modem_context *mctx, struct modem_data *mdata)
+{
+	int counter = 0, ret = -1;
+
+	while (counter < 50 && ret < 0) {
+
+		/* Send "AT" command to the modem. */
+		ret = modem_cmd_send(&mctx->iface, &mctx->cmd_handler,
+				     NULL, 0, "AT", &mdata->sem_response,
+				     MDM_CMD_TIMEOUT);
+
+		/* Check the response from the Modem. */
+		if (ret < 0 && ret != -ETIMEDOUT) {
+			return ret;
+		}
+
+		counter++;
+		k_sleep(K_SECONDS(2));
```

Comment:
I think the problem is not that it will wait 100 seconds for the modem to respond to `AT` commands, but that it will basically hang the system for this time. May be moving this code from the main thread to its own worker thread is a better solution ? Once the modem is initialized (in the background), it can send an event to application so that the app can start using this network device. I think PPP already works this way, but not sure. 

In any case, I think I will want to do a separate PR to fix this problem, if we agree to a solution.